neo/Ledger/MemoryPool.cs

```diff
@@ -4,162 +4,504 @@
+        {
+            _txRwLock.EnterWriteLock();
+            try
+            {
+                // First remove the transactions verified in the block.
+                foreach (Transaction tx in block.Transactions)
                 {
-                    RemoveLowestFee(_mem_pool_free, exceed);
-                    exceed = Count - Capacity;
+                    if (TryRemoveVerified(tx.Hash, out _)) continue;
+                    TryRemoveUnVerified(tx.Hash, out _);
+                }
 
-                    if (exceed > 0)
-                    {
-                        RemoveLowestFee(_mem_pool_fee, exceed);
-                    }
+                // Add all the previously verified transactions back to the unverified transactions
+                foreach (PoolItem item in _sortedHighPrioTransactions)
+                {
+                    if (_unverifiedTransactions.TryAdd(item.Transaction.Hash, item))
+                        _unverifiedSortedHighPriorityTransactions.Add(item);
+                }
+
+                // NOTE: This really shouldn't be necessary any more and can actually be counter-productive, since
+                //       they can be re-added anyway and would incur a potential addtional verification; setting to 1000
+                //       blocks for now. If transactions want to only be valid for short times they can include a script.
+                var lowPriorityCutOffTime = DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 1000);
+                foreach (PoolItem item in _sortedLowPrioTransactions)
+                {
+                    // Expire old free transactions
+                    if (item.Transaction.IsLowPriority && item.Timestamp < lowPriorityCutOffTime) continue;
+
+                    if (_unverifiedTransactions.TryAdd(item.Transaction.Hash, item))
+                        _unverifiedSortedLowPriorityTransactions.Add(item);
                 }
+
+                // Clear the verified transactions now, since they all must be reverified.
+                _unsortedTransactions.Clear();
+                _sortedHighPrioTransactions.Clear();
+                _sortedLowPrioTransactions.Clear();
             }
+            finally
+            {
+                _txRwLock.ExitWriteLock();
+            }
+
+            // If we know about headers of future blocks, no point in verifying transactions from the unverified tx pool
+            // until we get caught up.
+            if (block.Index < Blockchain.Singleton.HeaderHeight)
+                return;
+
+            if (Plugin.Policies.Count == 0)
+                return;
 
-            return pool.ContainsKey(hash);
+            LoadMaxTxLimitsFromPolicyPlugins();
+
+            ReverifyTransactions(_sortedHighPrioTransactions, _unverifiedSortedHighPriorityTransactions,
+                _maxTxPerBlock, MaxSecondsToReverifyHighPrioTx, snapshot);
+            ReverifyTransactions(_sortedLowPrioTransactions, _unverifiedSortedLowPriorityTransactions,
```

Comment:
If the node does not have any policy plugins installed, then it will get `int.MaxValue`, which will have to reverify all transactions after block persistence.

In fact, for non-consensus nodes, they do not install policy plugins, and there is no need to reverify transactions after block persistence. In this case, the reverification of the transaction can be handled in the `Idea` message.

Reply:
I already accounted for this, see line 494:
```
           if (Plugin.Policies.Count == 0)
                return;
```